`/signin` エンドポイントのpostメソッドに関する詳細設計を追加

a5ob7r opened the pull request:
# 概要

`/signin` エンドポイントのPOSTメソッドに関する詳細設計をOpenAPI v3.1.0に沿って記述した。

リクエストやレスポンスのパラメータはOAuth 2.0のRFCを参考に記述した。

エラーレスポンスは、Problem Detailsを参考に記述した。あくまで参考なので、Problem Detailsでは必須になっているパラメータを省略していたりする。

# 関連知識

- OpenAPI Specification v3.1.0 - https://spec.openapis.org/oas/v3.1.0
- JSON Schema - https://json-schema.org
- The OAuth 2.0 Authorization Framework - https://datatracker.ietf.org/doc/html/rfc6749

# 補足

本プルリクは #3 の後続である。 #3 のマージ後にベースブランチをmainに変更する。

a5ob7r committed:
Add an API description for `/signin`'s post

tk0miya commented on openapi.yaml:
```diff
@@ -41,7 +41,103 @@ components:
+                  enum:
+                    - password
+                  description: 認可サーバに要求するグラントのタイプである。ここではResource Owner Password Credential grant方式を利用するため、値は `password` である必要がある。
+                user_name:
```
RFC6749 に従うのであれば `user_name` ではなく `username` では?

a5ob7r replied:
あれ、本当だ。修正します。

tk0miya commented on openapi.yaml:
```diff
@@ -41,7 +41,103 @@ components:
+      operationId: LoginUser
+      requestBody:
+        description: |
+          OAuth2.0のResource Owner Password Credential grant方式によるアクセストークンリクエスト時に必要となるパラメータを格納する。
```
雑談: こんな記事を見かけました。
https://zenn.dev/ritou/articles/f88cbb888cb5c9142e74

念のためですが、今回はトレーニング用途なので、凝りすぎても疲れてしまうし、変える必要は感じていません。

a5ob7r replied:
一応補足しておくと、OAuth 2.0のRFCにも、基本的にResource Owner Password Credentialsグラントは利用するべきではない、といった旨の説明はありました。なので、現状の設計がよろしくないことは承知しています。本来は必要なクライアント認証もやっていませんし。

a5ob7r replied:
Resource Owner Password Credentials、略して `ROPC` か。

KT-tim commented on openapi.yaml:
```diff
@@ -41,7 +41,103 @@ components:
   /signin:
     post:
-      summary: ユーザとしてログインする
+      summary: ログインしたユーザにアクセストークンを発行する。
```
ログインするAPIじゃなくて、事前にログインした人向けに何かするAPIになってるような説明に見えるかな

a5ob7r replied:
`認証したユーザにアクセストークンを発行する。` だとどうでしょうか。

KT-tim replied:
その説明はあってると思う　ログイン、ログアウトって言葉の定義が意外と曖昧だから、使わない方が正確な説明になりそう　あるいはログイン、ログアウトという言葉の定義を明確にどこかに書くか

a5ob7r replied:
> `認証したユーザにアクセストークンを発行する。` だとどうでしょうか。

では、このように修正します。

a5ob7r replied:
[5a2509e](https://github.com/sls-training/2023-API-training-server/pull/4/commits/5a2509ec29d46aebd6aeaddec2733d33ce3e8356) にて修正しました。

KT-tim replied:
修正ナイス

> ユーザとしてログアウトする

下にこういうの書いてあったけどここも書き換える？（やるとしたら別プルリクでOK）

a5ob7r replied:
ちょうど #5 で書き換えてますね。

> 発行したアクセストークンを失効させる。

tk0miya commented on openapi.yaml:
```diff
@@ -41,7 +41,103 @@ components:
+                    description: アクセストークンが発行されてから失効するまでの有効期間を示す。単位は秒である。
+                  refresh_token:
+                    type: string
+                    description: 発行されたアクセストークンと同じグラントを利用して新しいアクセストークンを発行するためのトークンである。
```
今回 refresh_token まで取り扱います? optional なのでサボってもよさそう。

a5ob7r replied:
そういえばそうですね。勢いで書いてしまいましたが、API設計の本質とは少々ずれていますかね。それではリフレッシュトークンはサポートしないこととします。

tk0miya commented on openapi.yaml:
```diff
@@ -41,7 +41,103 @@ components:
+
+            https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
+          content:
+            application/json:
```
こっちも Problem Details 対応する?

a5ob7r replied:
ここは迷いどころでして、OAuth 2.0のRFCには、トークン発行の失敗時のレスポンスについての記載があり、そこではレスポンスボディには `application/json` を使用するとありました。なのでここでは Problem Details を利用しませんでした。

まあ、 メディアタイプが `application/json` か `application/problem+json` かの違いだし、プロパティ名も特に競合していないので割り切って後者にするのでもいいのかな、とも思っています。

大きな差がないとはいえ、エンドポイントごとにエラーレスポンスのメディアタイプが異なるのは、利用者からすればややこしいだろうなとは思っていました。

tk0miya replied:
クライアントにもよるのだけど、多くの API クライアントは content-type を厳密にチェックしないと思うので、Problem Details にしても (`application/problem+json`) にしても、別に構わないと思うよ。
※ ブラウザは厳密なチェックをするので要注意です

コメントしてもらったように、エンドポイントごとに振る舞いが変わる方がクライアントによって厄介なので、統一されていたほうがいいね。

>ここは迷いどころでして

迷ったところは相談してもらえるとお得です。
自分が決めかねている部分であれば、検討中やレビューの段階で相談しないと、解決するタイミングがないですよね。僕だったら PR の該当行に「ここは迷ったので意見もらえますか」ってコメントをつけると思う。

また、自分の中で結論をつけた場合についてもなにか補足があったほうがいいかもね。
自分も迷ったということは、レビュアーも迷うということで、不要なコミュニケーションが発生するおそれがあります。

最終的によいものがつくられないと、一旦マージしたとしても作り直しになってしまうので、小さいことであっても「これでいいのかな?」「ちょっと不安」というのはチームですり合わせておいたほうが無難です。

a5ob7r replied:
> クライアントにもよるのだけど、多くの API クライアントは content-type を厳密にチェックしないと思うので、Problem Details にしても (application/problem+json) にしても、別に構わないと思うよ。
※ ブラウザは厳密なチェックをするので要注意です
>
>コメントしてもらったように、エンドポイントごとに振る舞いが変わる方がクライアントによって厄介なので、統一されていたほうがいいね。

承知しました。では、 `application/problem+json` に統一します。

a5ob7r replied:
> 迷ったところは相談してもらえるとお得です。
> 自分が決めかねている部分であれば、検討中やレビューの段階で相談しないと、解決するタイミングがないですよね。僕だったら PR の該当行に「ここは迷ったので意見もらえますか」ってコメントをつけると思う。
>
> また、自分の中で結論をつけた場合についてもなにか補足があったほうがいいかもね。
> 自分も迷ったということは、レビュアーも迷うということで、不要なコミュニケーションが発生するおそれがあります。
>
> 最終的によいものがつくられないと、一旦マージしたとしても作り直しになってしまうので、小さいことであっても「これでいいのかな?」「ちょっと不安」というのはチームですり合わせておいたほうが無難です。

この辺りについては、すでに何度も指摘していただいている気がしますね。意識してみます。

tk0miya replied:
個人的な感覚としては、(なにかトラブルが起きたときに)「以前からその部分は気になっていたんですが…」と後出しされるのがチームの信頼を失う行為だと思っています。気になっていたのならなんで言わないの? と周りをしらけさせてしまう遠因となります。
※ 前もって伝えていたとしても、解決できるか、手を打っていたかはわからないのだけど、後出しされてしまうと個人の判断になってしまうので、トラブルの苛立ちが当人に集中してしまう可能性がある (本来はトラブルにぶつけないといけない)
※ 手を打たないにしても、個人の判断ではなく、チームの判断として深追いしないというのを決められるとよいですね

実装段階、設計段階で一瞬頭をよぎったものなどもあるだろうし、すべての気になったことをちゃんとチームに共有するというのは難しいとは思うのですが(熱中しているとつい忘れちゃうしw)、覚えているものはなるべく伝えておくのがよいとおもいます。

a5ob7r replied:
[a992bf8](https://github.com/sls-training/2023-API-training-server/pull/4/commits/a992bf8f59bd66b05d38184f465b7e8285ba2f67) にて、エラーレスポンスを Problem Details 対応させました。

a5ob7r committed:
Fix a property name in accordance with the RFC

https://datatracker.ietf.org/doc/html/rfc6749#section-4.3.2

a5ob7r committed:
Remove a refresh token support

This is optional according to the RFC, and probably the support isn't
related to API construction nature.

tk0miya commented on openapi.yaml:
```diff
@@ -41,7 +41,99 @@ components:
+                    type: string
+                    description: 失敗した理由に関する対人可読な追加情報である。
+                required:
+                  - erorr
```
typo 発見

a5ob7r replied:
あー、他のPRでも指摘されましたね。ここでタイポして、それをコピペしたということかな。

[076bc56](https://github.com/sls-training/2023-API-training-server/pull/4/commits/076bc56aef4f708bbc3a64f9cbaf12aded98a154) にて修正しました。

tk0miya commented on openapi.yaml:
```diff
@@ -41,7 +41,99 @@ components:
+                  error:
+                    type: string
+                    enum:
+                      - invalid_request
+                      - invalid_client
+                      - invalid_grant
+                      - unauthorized_client
+                      - unsupported_grant_type
```
他の PR ではこの enum 値、減らしてましたよね

a5ob7r replied:
[0b48488](https://github.com/sls-training/2023-API-training-server/pull/4/commits/0b484886240bda85aa291339346fe28e593a75c2) にて、発生し得るエラーコードのみになるよう修正しました。

tk0miya reviewed (approved):
コメントを付記しましたが、細かい内容なのでフライングで +1 します。

tk0miya reviewed (approved):
コメントを付記しましたが、細かい内容なのでフライングで +1 します。

a5ob7r committed:
Fix the summary for the `/signin`'s post method

a5ob7r committed:
Fix a typo

a5ob7r committed:
Fix available error codes

a5ob7r committed:
Add some schema components for Problem Details

a5ob7r committed:
Add some fields to an error response to adapt Problem Details
